driver/pelican/types/scheduleSet.go

```diff
@@ -0,0 +1,88 @@
+package types
+
+import (
+	"encoding/xml"
+	"fmt"
+
+	"github.com/parnurzeal/gorequest"
+	rrule "github.com/teambition/rrule-go"
+)
+
+func (pel *Pelican) SetSchedule(newSchedule *ThermostatSchedule) error {
+	// Retrieve thermostat's latest schedule for extra block deletion purposes
+	currentSchedule, currentErr := pel.GetSchedule()
+	if currentErr != nil {
+		return fmt.Errorf("Error retrieving thermostat %v's current schedule: %v", pel.Name, currentErr)
+	}
+
+	for day, blocks := range newSchedule.DaySchedules {
```

Comment:
Is it always guaranteed that the new schedule will contain an entry for all days of the week?

We're assuming that the input schedule is exhaustive, meaning it explicitly lists all of the blocks we want set. But say the user doesn't want any schedule entries for Monday. In this case, would the new schedule contain no entry for Monday, in which case we should delete all existing blocks for that day?

It seems like this loop wouldn't cover that case.

Reply:
I see what you're saying. You're correct, I did assume that "exhaustive" meant that the new schedule contains an entry for all days of the week.

However, after taking a look and thinking about the current code, I think my loop in this case would be okay? Say a user entered [Monday: {Schedule}, Friday: {Schedule}] for the "newSchedule *ThermostatSchedule" parameter, so all the days except for Monday and Friday are omitted.

My outermost for-loop iterates through the (key, value) pairs in the ThermostatSchedule object, which would just be "Monday: {Schedule}" and "Friday: {Schedule}" in this case, deleting and re-entering blocks for only those days. I don't think I use magic numbers like 7 days in a week, or hard code deletions for each day of the week. All my deletions and insertions are contained within the outer "DaySchedules" for loop. The first inner for loop clears all the existing blocks, and the second adds the new blocks.

I tested it with a dummy thermostat schedule and it seemed to work properly? I very likely am missing something, but from an initial glance I feel like it's okay?

Reply:
Say that the thermostat has an existing schedule with events on Monday, Wednesday, and Friday. Next, we get a request to set the thermostat's schedule with events specified only for Monday and Friday, like your example above.

Your loop would clear out the existing events for Monday and Friday, but leave Wednesday untouched, right? But if we are assuming that the specified schedule is exhaustive, then you would need to delete the existing schedule for Wednesday as well, which is not a key in `newSchedule.DaySchedules`.

Is my understanding correct or am I missing something where events for days not included in the new schedule are cleared out?

Reply:
Ah, I see what you're saying. Yes, that's absolutely correct, in the example above, I would fail to remove Wednesday.

It seems like what you're saying is that in the case I'm given an exhaustive schedule that only has two days' schedules specified, I should be removing the days that are not included? To make sure  I'm on the same page, that means if I have a Pelican schedule with schedules for all days Monday thru Sunday, and then receive a set request containing schedules for just Monday and Friday, I should remove the schedule of all the other days that are not included?

Reply:
Yes, that's right. This is because our simplifying assumption, for now, is that anyone who wishes to modify a Pelican's schedule has to specify the full schedule for the complete week, as opposed to just the days they want changed. So if the user omits a day of the week, that implies they don't want any events scheduled for that day.

Reply:
Got it, I will make sure to implement changes to reflect that! I think the most likely approach I'll take is just record which days were changed by the for loops, and then afterwards, subtract that subset from a list of days of the week to figure out which day schedules should be completely deleted.